### ISSUE TITLE
Failure to upgrade from EOL'd release

### ISSUE THREAD

**Issue by aral** [pop-os/upgrade#81]
Following instructions as per https://support.system76.com/articles/upgrade-pop/#advanced-install-terminal:

```sh
pop-upgrade release upgrade systemd
```

fails due to missing `sudo` with:

```sh
/sys/firmware/dmi/tables/smbios_entry_point: Permission denied
/dev/mem: Permission denied
Event: removing deprecated and/or conflicting packages
Event: updating package lists for the current release
Release upgrade status: release upgrade aborted: failed to update package lists for the current release: command failed with exit status: exit code: 100
pop-upgrade: dbus client error: calling ReleaseUpgradeFinalize method failed
```

Running with sudo afterwards fails with:

```sh
Event: removing deprecated and/or conflicting packages
Event: updating package lists for the current release
Release upgrade status: release upgrade aborted: failed to update package lists for the current release: command failed with exit status: exit code: 100
pop-upgrade: dbus client error: calling ReleaseUpgradeFinalize method failed
```

In GNOME Settings it shows Ubuntu 19.04. I can see the Download button. Pressing it results in the same Error 100.

It also seems my `sources.list` has been messed up.

Output of `sudo apt update`:

```sh
Hit:1 http://packages.microsoft.com/repos/vscode stable InRelease
Hit:2 http://ppa.launchpad.net/linuxuprising/shutter/ubuntu disco InRelease                                     
Ign:3 http://dl.google.com/linux/chrome/deb stable InRelease                                                    
Ign:4 http://old-releases.ubuntu.com/ubuntu disco InRelease                                                                   
Hit:5 http://dl.google.com/linux/chrome/deb stable Release                                                                    
Ign:6 http://old-releases.ubuntu.com/ubuntu disco-updates InRelease                                                           
Hit:7 http://ppa.launchpad.net/obsproject/obs-studio/ubuntu disco InRelease                                                   
Ign:8 http://old-releases.ubuntu.com/ubuntu disco-security InRelease                     
Ign:9 http://old-releases.ubuntu.com/ubuntu disco-backports InRelease              
Hit:10 http://ppa.launchpad.net/samoilov-lex/sway/ubuntu disco InRelease
Err:11 http://old-releases.ubuntu.com/ubuntu disco Release          
  404  Not Found [IP: 2001:67c:1360:8001::25 80]
Err:13 http://old-releases.ubuntu.com/ubuntu disco-updates Release                 
  404  Not Found [IP: 2001:67c:1360:8001::25 80]
Hit:14 http://ppa.launchpad.net/system76/pop/ubuntu disco InRelease 
Get:15 https://gitlab.com/paulcarroty/vscodium-deb-rpm-repo/raw/repos/debs vscodium InRelease [3,828 B]  
Err:16 http://old-releases.ubuntu.com/ubuntu disco-security Release                                                    
  404  Not Found [IP: 2001:67c:1360:8001::25 80]
Hit:17 http://ppa.launchpad.net/wireguard/wireguard/ubuntu disco InRelease         
Err:18 http://old-releases.ubuntu.com/ubuntu disco-backports Release
  404  Not Found [IP: 2001:67c:1360:8001::25 80]
Reading package lists... Done
E: The repository 'http://old-releases.ubuntu.com/ubuntu disco Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: The repository 'http://old-releases.ubuntu.com/ubuntu disco-updates Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: The repository 'http://old-releases.ubuntu.com/ubuntu disco-security Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: The repository 'http://old-releases.ubuntu.com/ubuntu disco-backports Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
```
and `cat /etc/apt/sources.list`:

```sh
# deb cdrom:[Pop_OS 18.04 _Bionic Beaver_ - Release amd64 (20180612)]/ bionic main restricted
deb-src http://old-releases.ubuntu.com/ubuntu/ disco main restricted universe multiverse
deb http://old-releases.ubuntu.com/ubuntu/ disco-updates main restricted universe multiverse
deb-src http://old-releases.ubuntu.com/ubuntu/ disco-updates main restricted universe multiverse
deb http://old-releases.ubuntu.com/ubuntu/ disco-security main restricted universe multiverse
deb-src http://old-releases.ubuntu.com/ubuntu/ disco-security main restricted universe multiverse
deb http://old-releases.ubuntu.com/ubuntu/ disco-backports main restricted universe multiverse
deb-src http://old-releases.ubuntu.com/ubuntu/ disco-backports main restricted universe multiverse
deb http://old-releases.ubuntu.com/ubuntu/ disco main restricted universe multiverse
# deb http://apt.pop-os.org/proprietary disco main
# deb-src http://apt.pop-os.org/proprietary bionic main

deb https://gitlab.com/paulcarroty/vscodium-deb-rpm-repo/raw/repos/debs/ vscodium main
```

**Comment by aral** [comment]
OK, this is how I fixed it:

Updated `/etc/apt/sources.list` file to:

```
deb http://de.archive.ubuntu.com/ubuntu disco main restricted universe multiverse
deb http://de.archive.ubuntu.com/ubuntu disco-updates main restricted universe multiverse
deb http://de.archive.ubuntu.com/ubuntu disco-security main restricted universe multiverse
deb http://de.archive.ubuntu.com/ubuntu disco-backports main restricted universe multiverse
```

```sh
sudo apt update
sudo apt install pop-desktop sessioninstaller
```

Then replaced `/etc/apt/sources.list` with:

```
deb http://de.archive.ubuntu.com/ubuntu eoan main restricted universe multiverse
deb http://de.archive.ubuntu.com/ubuntu eoan-updates main restricted universe multiverse
deb http://de.archive.ubuntu.com/ubuntu eoan-security main restricted universe multiverse
deb http://de.archive.ubuntu.com/ubuntu eoan-backports main restricted universe multiverse
```

Then used the Download button in Settings.

**Comment by aral** [comment]
(Also not sure if its related but I purged gnome-software as well.)

Note that even with the above, `pop-upgrade release upgrade systemd` was giving me the same error but the Download button in Settings appears to be running (it’s at “preparing upgrade”).

**Comment by aral** [comment]
Hmm, maybe I spoke too soon. Current status:

![Screenshot from 2020-01-20 19-27-18](https://user-images.githubusercontent.com/41798/72752769-f8106000-3bba-11ea-8503-c8093bbed708.png)

Going to try rebooting and see what that does :crossed_fingers:

**Comment by aral** [comment]
Right, I didn’t reboot. Output of what I tried, now I’m going to reboot:

```sh
$ sudo pop-upgrade release upgrade systemd
pop-upgrade: dbus client error: calling ReleaseCheck method failed
 $ sudo apt full-upgrade
E: dpkg was interrupted, you must manually run 'sudo dpkg --configure -a' to correct the problem. 
$ sudo dpkg --configure -a
Processing triggers for libc-bin (2.30-0ubuntu2) ...
Processing triggers for man-db (2.8.7-3) ...
Processing triggers for hicolor-icon-theme (0.17-2) ...
$ sudo apt full-upgrade
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Calculating upgrade... Done
The following packages have been kept back:
  libomp-dev
0 to upgrade, 0 to newly install, 0 to remove and 1 not to upgrade.
1 not fully installed or removed.
After this operation, 0 B of additional disk space will be used.
Do you want to continue? [Y/n] 
Setting up pop-upgrade (0.1.0~1579032976~19.04~e90d1d0) ...
+ systemctl --global enable pop-upgrade-notify.timer
Failed to enable unit, unit pop-upgrade-notify.timer does not exist.
+ [ configure = configure ]
+ deb-systemd-helper unmask pop-upgrade.service
+ deb-systemd-helper --quiet was-enabled pop-upgrade.service
+ deb-systemd-helper enable pop-upgrade.service
+ [ configure = configure ]
+ [ -d /run/systemd/system ]
+ systemctl --system daemon-reload
+ deb-systemd-invoke start pop-upgrade-init.service pop-upgrade.service
pop-upgrade-init.service is a disabled or a static unit, not starting it.
+ exit 0
$ sudo pop-upgrade release upgrade systemd
no release available to upgrade to
```

**Comment by aral** [comment]
Great, rebooted and GNOME's gone. Writing this from the Awesome display manager I'd thankfully installed a while back :(

**Comment by mmstick** [comment]
The daemon needs to be fixed to only switch Ubuntu archives to `old-releases` if that release has actually moved to `old-releases` yet.

**Comment by mmstick** [comment]
In the future, I would recommend always upgrading before the EOL date is reached, since that's when all support for that release ends.

**Comment by aral** [comment]
Hiya @mmstick, thanks for the quick PR :)

I’m trying to build this and tried both make and cargo and I’m getting (for cargo check): failed to run custom build command for libparted-sys v0.1.1 … parter/parted.h file not found

Any ideas?

(I managed to downgrade my kernel and install pop-desktop. Release version of pop-upgrade now reports 19.10 as available. Hoping running your EOL branch will fix things.)

**Comment by mmstick** [comment]
Run `sudo apt build-dep pop-upgrade` to fetch build dependencies for debian packages.

You will need to revert your apt sources back to `us.archive`, which have been switched to `old-releases`.

**Comment by aral** [comment]
Thanks, @mmstick, running that, followed by `sudo apt update` and `sudo apt full-upgrade` has got me a functional Pop!_OS 19.04 downgrade. Also, it got me past the libparted error in building the upgrade app, now make aborts with:

```sh
error[E0425]: cannot find function `head` in crate `reqwest`
   --> src/repair/sources.rs:145:14
    |
145 |     reqwest::head(&release_file).send().ok().map_or(false, |resp| resp.status().is_success())
    |              ^^^^ not found in `reqwest`

warning: unused `#[macro_use]` import
  --> src/lib.rs:13:1
   |
13 | #[macro_use]
   | ^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused import: `reqwest::Client`
 --> src/repair/sources.rs:6:5
  |
6 | use reqwest::Client;
  |     ^^^^^^^^^^^^^^^

error: aborting due to previous error
```

**Comment by aral** [comment]
Is there an ETA on when this updated pop-upgrade package will show up in apt? I can’t compile it due to the error about (I’m assuming there is an undocumented dependency not handled by Cargo/Make?)

My setup is in limbo right now. Would love to be able to run the upgrade. Any ideas? Thanks in advance!

**Comment by brs17** [comment]
> Is there an ETA on when this updated pop-upgrade package will show up in apt? I can’t compile it due to the error above (I’m assuming there is an undocumented dependency not handled by Cargo/Make?)
> 
> My setup is in limbo right now. Would love to be able to run the upgrade. Any ideas? Thanks in advance!

I just released it within the past hour. It is building in launchpad and then will be available soon thereafter.

**Comment by aral** [comment]
@brs17 Just saw it; thanks so much! About to give it a shot :crossed_fingers:

**Comment by aral** [comment]
w00t, cooking on Pop!_OS 19.10 juice. Thanks for your rapid response on this :) :two_hearts: 

(The upgrade initially failed while trying to install pop-desktop but I was able to recover from Terminal using aptitude to track down the culprit package – nautilus – that was blocking the installation and downgrading it.)

![Screenshot from 2020-01-21 14-50-14](https://user-images.githubusercontent.com/41798/72815039-f21f8b00-3c5d-11ea-8feb-8315270fb51c.png)

![Screenshot from 2020-01-21 14-51-54](https://user-images.githubusercontent.com/41798/72815082-ffd51080-3c5d-11ea-8bfa-e4305b774621.png)

**Comment by aral** [comment]
(Oh wait, it still says 19.04. Going to try a reboot.)

**Comment by aral** [comment]
OK, so I managed to get my system back to pristine 19.04 and upgrade to 19.10 by going back to the tried and true:

```sh
sudo apt update
sudo apt full-upgrade
sudo pop-upgrade release upgrade systemd
```

(The full-upgrade downgraded a bunch of packages to disco.)

And tadaaaaa!

Thanks again for your help :) The actual upgrade process (once the bug was fixed) was super smooth. Great job on the upgrader (although I still don’t understand why I can’t compile it from source). 

![Screenshot from 2020-01-21 15-14-35](https://user-images.githubusercontent.com/41798/72817182-45dfa380-3c61-11ea-878c-55471f79c578.png)
